bigtable/metrics.go

```diff
@@ -0,0 +1,487 @@
+				return
+			}
+
+			// Calculate elapsed time
+			elapsedTime := float64(time.Since(startTime).Nanoseconds()) / 1000000
+
+			// Attributes for attempt_latencies
+			attemptLatCurrCallAttrs, attemptLatErr := mt.toOtelMetricAttrs(metricNameAttemptLatencies)
+			attemptLatAllAttrs := metric.WithAttributes(append(mt.clientAttributes, attemptLatCurrCallAttrs...)...)
+
+			// Attributes for server_latencies
+			serverLatCurrCallAttrs, serverLatErr := mt.toOtelMetricAttrs(metricNameServerLatencies)
+			serverLatAllAttres := metric.WithAttributes(append(mt.clientAttributes, serverLatCurrCallAttrs...)...)
+
+			if attemptLatErr == nil {
+				mt.instrumentAttemptLatencies.Record(mt.ctx, elapsedTime, attemptLatAllAttrs)
+			}
+
+			if serverLatErr == nil {
+				serverLatency, serverLatErr := mt.getServerLatency()
+				if serverLatErr == nil {
+					mt.instrumentServerLatencies.Record(mt.ctx, serverLatency, serverLatAllAttres)
+				}
+			}
+		}()
+
+		// Make call to CBT service
+		err := f(ctx, callSettings)
+
+		// Record attempt status
+		statusCode, _ := convertToGrpcStatusErr(err)
+		mt.status = statusCode.String()
+		return err
+	}
+	return gax.Invoke(ctx, callWrapper, opts...)
+}
+
+func (mt *builtinMetricsTracer) recordAndConvertErr(err error) error {
+	statusCode, statusErr := convertToGrpcStatusErr(err)
+	mt.status = statusCode.String()
+	return statusErr
+}
+
+// toOtelMetricAttrs converts recorded metric attributes values to OpenTelemetry attributes format
+func (mt *builtinMetricsTracer) toOtelMetricAttrs(metricName string) ([]attribute.KeyValue, error) {
+	clusterID, zoneID, _ := mt.getLocation()
+
+	// Create attribute key value pairs for attributes common to all metricss
+	attrKeyValues := []attribute.KeyValue{
+		attribute.String(metricLabelKeyMethod, mt.method),
+
+		// Add resource labels to otel metric labels.
+		// These will be used for creating the monitored resource but exporter
+		// will not add them to Google Cloud Monitoring metric labels
+		attribute.String(monitoredResLabelKeyTable, mt.tableName),
+		attribute.String(monitoredResLabelKeyCluster, clusterID),
+		attribute.String(monitoredResLabelKeyZone, zoneID),
+	}
+
+	attrs, found := additionalAttributes[metricName]
+	if !found {
+		return nil, fmt.Errorf("Unable to create attributes list for unknown metric: %v", metricName)
+	}
+
+	// Add additional attributes to metrics
+	for _, attrKey := range attrs {
+		switch attrKey {
+		case metricLabelKeyOperationStatus:
+			attrKeyValues = append(attrKeyValues, attribute.String(metricLabelKeyOperationStatus, mt.status))
+		case metricLabelKeyStreamingOperation:
+			attrKeyValues = append(attrKeyValues, attribute.Bool(metricLabelKeyStreamingOperation, mt.isStreaming))
+		default:
+			return nil, fmt.Errorf("Unknown additional attribute: %v", attrKey)
+		}
+	}
+
+	return attrKeyValues, nil
+}
+
+// get GFE latency from response metadata
+func (mt *builtinMetricsTracer) getServerLatency() (float64, error) {
```

Comment:
Similar to above...I dont think the tracer should know about the encoding of headers...it should be solely responsible for being told what the metric values are and it should be either the client or another utility that extracts those values

Reply:
Done